doc/GUIDE.md

```diff
@@ -0,0 +1,848 @@
+
+```c
+// mylib.h
+
+void foo_variadic(int x, ...);
+```
+
+Jextract doesn't generates a regular method, but a _class_, which represents the invoker:
+
+```java
+// mylib.h
+
+public static class foo_variadic {
+    public static MemorySegment address() { ... }
+    public static foo_variadic makeInvoker(MemoryLayout... layouts) { ... }
+
+    // not static!
+    public MethodHandle handle() { ... }
+    public FunctionDescriptor descriptor() { ... }
+    public void apply(int x, Object... x1) { ... }
+}
+```
+
+This class has a `static` meta-data accessor for the function address, and a `makeInvoker`
+factory method which can be used to create an instance on the invoker class. The `MemoryLayout...`
+arguments passed to `makeInvoker` represent the memory layouts of the variadic parameters
+that are to be passed to the function. The `makeInvoker` factory essentially instantiates
+the variadic function (like you would a template) for a particular set of parameter types.
+
+We can then use the instance methods `apply` or `handle` if we want to invoke the function,
+as follows:
+
+```java
+// Main.java
+
+foo_variadic invoker = foo_variadic.makeInvoker(C_INT, C_INT, C_INT);
+invoker.apply(3, 1, 2, 3);
+invoker.handle().invokeExact(3, 1, 2, 3);
+```
+
+Here we instantiate `foo_variadic` for 3 parameter types of the C type `int`. These parameter
+types essentially replace the `...` ellipsis in the C function type.
+
+We can call the instantiated invoker either by calling `apply`, which will box the arguments
+into an `Object[]`, or through the method handle returned by `handle()` which avoids the
+overhead of boxing.
+
+### Typedefs
+
+As mentioned before: typedefs are either translated as a `static final` memory layout fields
+in the main header class that jextract generates, or as a separate class, depending on
+whether the typedef is for a primitive type or a struct/union type respectively.
+
+Take for example the following `typedef` declarations:
+
+```c
+// mylib.h
+
+typedef int MyInt;
+
+struct Point {
+    int x;
+    int y;
+};
+typedef struct Point MyPoint;
+```
+
+`MyInt` is a `typedef` of the primitive type `int`, so it is translated by jextract as a
+`static final` layout field in the main header class:
+
+```java
+// mylib_h.java
+
+public static final OfInt MyInt = mylib_h.C_INT;
+```
+
+The `MyPoint` `typedef` on the other hand is a typedef for a struct, so it is translated
+as a separate class which extends the class that is generated for the `Point` struct:
+
+```java
+// MyPoint.java
+
+public class MyPoint extends Point { }
+```
+
+Through static inheritance, all the methods in the `Point` class are available through the
+`MyPoint` class as well.
+
+### Nested Types
+
+C allows variable declarations to have an inline anonymous type. Jextract handles in
+particular cases where a struct's field has an inline type specially. For instance, if we
```

Comment:
"Jextract handles in particular cases where a struct's field has an inline type specially."

I don't understand this sentence. Does Jextract handle nested types only when a struct's field has an inline type?

Reply:
C doesn't allow other kinds of nested types, so yeah

Reply:
I'll just remove the 'Jextract handles in particular cases where a struct's field has an inline type specially.' part here